src/cargo/util/toml/mod.rs

```diff
@@ -1578,10 +1578,25 @@ impl TomlManifest {
         let cargo_features = me.cargo_features.as_ref().unwrap_or(&empty);
         let features = Features::new(cargo_features, config, &mut warnings, source_id.is_path())?;
 
-        let project = me.project.clone().or_else(|| me.package.clone());
-        let project = &mut project.ok_or_else(|| anyhow!("no `package` section found"))?;
+        let mut package = match (&me.package, &me.project) {
+            (Some(_), Some(_)) => {
+                bail!("cannot specify both `package` and `project` in the same `Cargo.toml`")
+            }
```

Comment:
It might be rare but some old projects would fail to compile after this PR, since the old behaviour favors `[project]` over `[package]` if co-exist. Could we keep the old behaviour for a while and emit a warning at least for a transition period?